exporter/xray/segment.go

```diff
@@ -0,0 +1,319 @@
+	//	* The time of the original request, in Unix epoch time, in 8 hexadecimal digits.
+	//	* For example, 10:00AM December 2nd, 2016 PST in epoch time is 1480615200 seconds, or 58406520 in hexadecimal.
+	//	* A 96-bit identifier for the trace, globally unique, in 24 hexadecimal digits.
+	TraceID string `json:"trace_id,omitempty"`
+
+	// EndTime - number that is the time the segment was closed. For example, 1480615200.090
+	// or 1.480615200090E9. Specify either an end_time or in_progress.
+	EndTime float64 `json:"end_time"`
+
+	/* ---------------------------------------------------- */
+
+	// Service - An object with information about your application.
+	//Service service `json:"service,omitempty"`
+
+	// User - A string that identifies the user who sent the request.
+	//User string `json:"user,omitempty"`
+
+	// Origin - The type of AWS resource running your application.
+	Origin string `json:"origin,omitempty"`
+
+	// ParentID – A subsegment ID you specify if the request originated from an instrumented
+	// application. The X-Ray SDK adds the parent subsegment ID to the tracing header for
+	// downstream HTTP calls.
+	ParentID string `json:"parent_id,omitempty"`
+
+	// Annotations - object with key-value pairs that you want X-Ray to index for search
+	Annotations map[string]interface{} `json:"annotations,omitempty"`
+
+	// SubSegments contains the list of child segments
+	SubSegments []*segment `json:"subsegments,omitempty"`
+
+	// Service - optional service definition
+	Service *service `json:"service,omitempty"`
+}
+
+type service struct {
+	// Name of the service4
+	Name string `json:"name,omitempty"`
+
+	// Version - A string that identifies the version of your application that served the request.
+	Version string `json:"version,omitempty"`
+}
+
+type http struct {
+}
+
+// makeID - returns a 64-bit identifier for the segment, unique among segments in
+// the same trace, in 16 hexadecimal digits.
+func makeID(spanID trace.SpanID) string {
+	if spanID == zeroSpanID {
+		return ""
+	}
+	return hex.EncodeToString(spanID[0:8])
+}
+
+// MakeTraceID – A unique identifier that connects all segments and subsegments
+// originating from a single client request.
+//  * A trace_id consists of three numbers separated by hyphens. For example,
+//    1-58406520-a006649127e371903a2de979. This includes:
+//  * The version number, that is, 1.
+//  * The time of the original request, in Unix epoch time, in 8 hexadecimal digits.
+//  * For example, 10:00AM December 2nd, 2016 PST in epoch time is 1480615200 seconds,
+//    or 58406520 in hexadecimal.
+//  * A 96-bit identifier for the trace, globally unique, in 24 hexadecimal digits.
+func MakeTraceID(traceID trace.TraceID, startTime time.Time) string {
+	const (
+		digits24 = "123456789012345678901234" // 24 digits
+	)
+
+	var (
+		content = make([]byte, 0, traceIDLength)
+		epoch   = startTime.Unix()
+	)
+
+	content = append(content, "1-"...)
+	content = strconv.AppendInt(content, epoch, 16)
```

Comment:
This assumes that startTIme is the start of the original request in the trace, but as this is called from `makeSegment`, this will have startTime = start time of each separate span. In general in OpenCensus you don't have access to the start time of the first span in the trace (this could be in a different process). I think what we need to do here is implement an X-Ray propagation.HTTPFormat for X-Ray that reads the header as described here:
https://docs.aws.amazon.com/xray/latest/devguide/xray-services-elb.html

The propagation.HTTPFormat will then create SpanContexts with TraceID suitable for using in X-Ray. This method then needs to convert from that TraceID to X-Ray's format. From that document above, it looks like the version 1 X-Ray format requires 16 bytes (encoded as 32 hex chars). So that should fit into a TraceID.

Reply:
Agreed for the most part, but there's a snag.  The first 8 chars are a unix epoch.  xray will check to ensure that the value is no older than 30 days and I believe there's a similar limit on how far in the future the value can be.

I can see a number of solutions for this and would love to get your thoughts:

* accept that traceIDs generated by oc will create new traces when propagated remotely
* allow oc to accept a custom traceID generator
* change the way oc generates traceIDs to be a 32bit timestamp plus 96bits of randomness

My preferences are for last, middle, and first solutions in that order.

Reply:
I agree with that order of preference :) I still think having an HTTPFormat would somewhat obviate the need for any self-generated TraceIDs. I assume ELB (or something) can generate them for you and you just read them off headers.